File: backend/src/main/java/com/carrental/backend/vehicle/VehicleService.java
package com.carrental.backend.vehicle;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class VehicleService {

    private final VehicleRepository vehicleRepository;

    private static final Set<String> ALLOWED_SORT_FIELDS = Set.of(
            "id",
            "brand",
            "model",
            "type",
            "driveType",
            "pricePerDay",
            "year"
    );

    public List<Vehicle> getVehicles(
            String type,
            String brand,
            String model,
            String driveType,
            Double minPrice,
            Double maxPrice,
            String sortBy,
            String sortDirection
    ) {
        Specification<Vehicle> spec = ((root, query, cb) -> cb.conjunction());
        if (type != null && !type.isBlank()) {
            spec = spec.and(((root, query, cb) ->
                    cb.equal(cb.lower(root.get("type")), type.toLowerCase())));
        }

        if (brand != null && !brand.isBlank()) {
            spec = spec.and((root, query, cb) ->
                    cb.equal(cb.lower(root.get("brand")), brand.toLowerCase()));
        }

        if (model != null && !model.isBlank()) {
            spec = spec.and((root, query, cb) ->
                    cb.like(cb.lower(root.get("model")), "%" + model.toLowerCase() + "%"));
        }

        if (driveType != null && !driveType.isBlank()) {
            spec = spec.and((root, query, cb) ->
                    cb.equal(cb.lower(root.get("driveType")), driveType.toLowerCase()));
        }

        if (minPrice != null) {
            spec = spec.and((root, query, cb) ->
                    cb.greaterThanOrEqualTo(root.get("pricePerDay"), minPrice));
        }

        if (maxPrice != null) {
            spec = spec.and((root, query, cb) ->
                    cb.lessThanOrEqualTo(root.get("pricePerDay"), maxPrice));
        }

        Sort.Direction direction = sortDirection.equalsIgnoreCase("desc")
                ? Sort.Direction.DESC
                : Sort.Direction.ASC;

        if (!ALLOWED_SORT_FIELDS.contains(sortBy)) {
            sortBy = "id";
        }

        Sort sort = Sort.by(direction, sortBy);

        return vehicleRepository.findAll(spec, sort);
    }

    public Vehicle getVehicle(Integer id) {
        return vehicleRepository.findById(id)
                .orElseThrow();
    }

    public Vehicle createVehicle(Vehicle vehicle) {
        return vehicleRepository.save(vehicle);
    }

    public void deleteVehicle(Integer id) {
        vehicleRepository.deleteById(id);
    }

    public Vehicle updateVehicle(Integer id, Vehicle updatedVehicle) {
        Vehicle vehicle = vehicleRepository.findById(id).orElseThrow();

        vehicle.setBrand(updatedVehicle.getBrand());
        vehicle.setModel(updatedVehicle.getModel());
        vehicle.setYear(updatedVehicle.getYear());
        vehicle.setType(updatedVehicle.getType());
        vehicle.setDriveType(updatedVehicle.getDriveType());
        vehicle.setPricePerDay(updatedVehicle.getPricePerDay());
        vehicle.setAvailable(updatedVehicle.isAvailable());
        vehicle.setDescription(updatedVehicle.getDescription());
        vehicle.setImageUrl(updatedVehicle.getImageUrl());

        return vehicleRepository.save(vehicle);
    }
}

File: frontend/src/services/vehicleService.ts
import api from './api';
import type { Vehicle, VehicleFilter } from '../types/Vehicle';

export const vehicleService = {
  async getVehicles(filters?: VehicleFilter): Promise<Vehicle[]> {
    const params = new URLSearchParams();
    if (filters?.type) params.append('type', filters.type);
    if (filters?.brand) params.append('brand', filters.brand);
    if (filters?.model) params.append('model', filters.model);
    if (filters?.driveType) params.append('driveType', filters.driveType);
    if (filters?.minPrice !== undefined) params.append('minPrice', filters.minPrice.toString());
    if (filters?.maxPrice !== undefined) params.append('maxPrice', filters.maxPrice.toString());
    if (filters?.sortBy) params.append('sortBy', filters.sortBy);
    if (filters?.sortDirection) params.append('sortDirection', filters.sortDirection);

    const { data } = await api.get<Vehicle[]>('/vehicles', { params });
    return data;
  },

  async getVehicle(id: number): Promise<Vehicle> {
    const { data } = await api.get<Vehicle>(`/vehicles/${id}`);
    return data;
  },

  async createVehicle(vehicle: Omit<Vehicle, 'id'>): Promise<Vehicle> {
    const { data } = await api.post<Vehicle>('/vehicles', vehicle);
    return data;
  },

  async updateVehicle(id: number, vehicle: Partial<Vehicle>): Promise<Vehicle> {
    const { data } = await api.put<Vehicle>(`/vehicles/${id}`, vehicle);
    return data;
  },

  async deleteVehicle(id: number): Promise<void> {
    await api.delete(`/vehicles/${id}`);
  },
};
